jlm/llvm/backend/rvsdg2dot/rvsdg2dot.cpp

```diff
@@ -0,0 +1,147 @@
+    typeGraph.CreateDirectedEdge(elementTypeNode, node);
+  }
+  // TODO: Function types are also aggregate types (arguments and results)
+
+  return node;
+}
+
+/**
+ * Fill the given \p graph with nodes corresponding to the nodes of the given \p region.
+ * If \p typeGraph is not nullptr, all rvsdg outputs get a type reference to the type graph.
+ * If the type does not already exist in the type graph, it is created.
+ */
+static void
+CreateGraphNodes(util::Graph & graph, rvsdg::region & region, util::Graph * typeGraph)
+{
+  graph.SetProgramObject(region);
+
+  // Connects an input port in the GraphWriter graph to an input in the RVSDG.
+  // Also adds an edge from the input's origin to the input port in the graph.
+  auto AttachInput = [&](util::Port & inputPort, const rvsdg::input & rvsdgInput)
+  {
+    inputPort.SetProgramObject(rvsdgInput);
+
+    if (auto originPort =
+            reinterpret_cast<util::Port *>(graph.GetElementFromProgramObject(*rvsdgInput.origin())))
+    {
+      auto & edge = graph.CreateDirectedEdge(*originPort, inputPort);
+      if (rvsdg::is<MemoryStateType>(rvsdgInput.type()))
+        edge.SetAttribute("color", util::Colors::Red);
+      if (rvsdg::is<iostatetype>(rvsdgInput.type()))
+        edge.SetAttribute("color", util::Colors::Green);
+    }
+  };
+
+  // Connects an output port in the GraphWriter graph to an output in the RVSDG
+  // Also adds a type attribute referencing a node in the typeGraph, if one exists.
+  auto AttachOutput = [&](util::Port & outputPort, const rvsdg::output & rvsdgOutput)
+  {
+    outputPort.SetProgramObject(rvsdgOutput);
+    if (typeGraph)
+      outputPort.SetAttributeGraphElement(
+          "type",
+          GetOrCreateTypeGraphNode(rvsdgOutput.type(), *typeGraph));
+  };
+
+  for (size_t n = 0; n < region.narguments(); n++)
+    AttachOutput(graph.CreateArgumentNode(), *region.argument(n));
+
+  rvsdg::topdown_traverser traverser(&region);
+  for (const auto rvsdgNode : traverser)
+  {
+    auto & node = graph.CreateInOutNode(rvsdgNode->ninputs(), rvsdgNode->noutputs());
+    node.SetLabel(rvsdgNode->operation().debug_string());
+    node.SetProgramObject(*rvsdgNode);
+
+    for (size_t i = 0; i < rvsdgNode->ninputs(); i++)
+      AttachInput(node.GetInputPort(i), *rvsdgNode->input(i));
+
+    for (size_t i = 0; i < rvsdgNode->noutputs(); i++)
+      AttachOutput(node.GetOutputPort(i), *rvsdgNode->output(i));
+
+    // Structural nodes also have subgraphs
+    if (auto structuralNode = dynamic_cast<const rvsdg::structural_node *>(rvsdgNode))
+    {
+      for (size_t i = 0; i < structuralNode->nsubregions(); i++)
+      {
+        auto & subGraph = node.CreateSubgraph();
+        CreateGraphNodes(subGraph, *structuralNode->subregion(i), typeGraph);
+        // FIXME: Attach the arguments and results from the subgraph to node's inputs and outputs
+        // Requires special logic for each type of structural node
```

Comment:
What would it take to fix this?

Reply:
Honestly it might be more straightforward than i initially thought, as structural results and arguments reference their structural inputs and outputs when applicable